### ISSUE TITLE
Race issue on duplicate key error

### ISSUE THREAD

**Issue by geosebas** [realm/realm-dart#1209]
### What happened?

Hello, 

I got a problem when trying inside the same transaction to first remove some objects and then replace them with new ones (but with same duplicate key) 

Here a very simplified example of what I'm trying to do : 
```dart
void example() async {  
  realm.write(() {
    var eventsToDelete = realm.query<SomeEvent>(r'groupId = $0', [oldEvent.groupId])
    realm.deleteMany(eventsToDelete);

    for(final newEvent in someList) {
      // I have a duplicate key on ['iso-date', 'slot']
      realm.add(SomeEvent('_id', 'iso-date', 'slot', 'groupId'));
    }
  })
  await realm.syncSession.waitForUpload();
}
```
I said 'race issue' in the title because, some new events will not have any problem, and some will. 
Just before writing this issue, I tried to delete 6 events, then created 6 new events with same duplicate key.
Of the 6, 4 got an error, and for 2 it was okay. 

If you need more context or anything do tell me. 

Thanks

### Repro steps

See example

### Version

Flutter 3.7.7

### What Atlas Services are you using?

Both Atlas Device Sync and Atlas App Services

### What type of application is this?

Flutter Application

### Client OS and version

Realm 1.02

**Comment by nirinchev** [comment]
What is the error you're getting? That is not clear from the issue description.

**Comment by geosebas** [comment]
Sorry here an example of the error I got in atlas : 

```
Error:

Failed to apply changes to MongoDB. This has been resolved by removing the object from the mobile device.

Source:

Error syncing MongoDB write

Logs:

[
  "MongoDB error: E11000 duplicate key error collection: padel-app-db.Match index: unique_match dup key: { accountId: \"xx\", courtId: \"d1a90a55-760b-48ef-aa7b-cc134f270c64\", date: \"2023-03-27\", slotIndex: 0 }",
  "Realm table: \"Match\"",
  "Realm object primary key: {2d6d70bb-c2a0-4e6d-a245-88faff0bd50a}"
]
```

**Comment by nielsenko** [comment]
This looks fishy:
```dart
realm.add(SomeEvent('_id',...
```
Why the string: `'_id'`?

**Comment by nielsenko** [comment]
You have an extra index server-side on the MongoDB collection called `unique_match` that is complaining about duplicate entries. This is illegal, and fixed by a compensating-write on the device.

**Comment by geosebas** [comment]
Sorry it was just to say which properties we're setting. 
Imagine a object that look like this : 
```dart
@RealmModel()
class _SomeEvent{
  @MapTo('_id')
  @PrimaryKey()
  late String id;
  late String groupId;
  late DateTime date;
  late int slot
}
```
Creating this object will look like this :
```dart
realm.add(SomeEvent('uuid', 'a group id', DateTime.now(), 5));
```

**Comment by geosebas** [comment]
> You have an extra index server-side on the MongoDB collection called unique_match that is complaining about duplicate entries. This is illegal, and fixed by a compensating-write on the device.

That's exactly my problem. Sorry if I was not clear.

**Comment by nielsenko** [comment]
What do you expect to happen here? 

You are creating an object that is invalid according to your index. We need to reconcile the data on the client and the backend. We do that by a so called compensating-write on the client, that removes the illegal object you created.

This is an asynchronous operation, can happen days later, if the device is offline. But even when online it is not immediate.

**Comment by nielsenko** [comment]
Are you sure you are not creating duplicates between the new objects?

**Comment by nielsenko** [comment]
I agree it should **not** be a problem to have a conflict with an object that is already dead.

**Comment by geosebas** [comment]
Imagine I have a composite unique index on `groupId`, `date`, and `slot` from the `SomeEvent` class (see previous comment)

Then here what I'm trying to do : 
```dart
final date = DateTime.now();
final event1 = SomeEvent('uuid1', 'groupID', date, 5);
realm.write(() => realm.add(event1));

// Some times later 

realm.write(() {
  realm.delete(event1);
  realm.add(SomeEvent('newUuid', 'groupID', date, 5));
});
```
The second write will cause an error in atlas because of my unique index, but I don't understand why since `event1` has been deleted before the adding the other one in the transaction

**Comment by geosebas** [comment]
> What do you expect to happen here?

The first event should be deleted, the second will be added without any conflits with my unique index

**Comment by nielsenko** [comment]
I agree that `event1` should never conflict with the new event, since it is deleted before the new event is created. However, are you sure some other client is not racing you?

**Comment by nielsenko** [comment]
I have reached out to a colleague. Will get back when I know more. Are you saying https://github.com/realm/realm-dart/issues/1209#issuecomment-1468318086 fails 100% of the times, or sometimes?

**Comment by geosebas** [comment]
> I have reached out to a colleague. Will get back when I know more. Are you saying [#1209 (comment)](https://github.com/realm/realm-dart/issues/1209#issuecomment-1468318086) fails 100% of the times, or sometimes?

Not 100% but at least 50% I would say. 

I will try to create a simple project to reproduce a bit later (in about 3/4 hours I think), so that you can test easily.

**Comment by geosebas** [comment]
@nielsenko here the project to reproduce : https://github.com/geosebas/flutter_realm_unique_index_issue

So far I managed to reproduce each time I tried.

**Comment by tkaye407** [comment]
Hi, thank you for helping to identify this issue. My name is Tyler and I work on the Device Sync Server team. This is an implementation detail of how we apply changes in Realm to MongoDB. What you are describing sounds like an issue on our side (interestingly one that has always existed but no-one has ever run into before). 

The mini description of the problem is that in the conversion from Realm instructions (each field is a single instruction) to MongoDB Writes, we do the following: 
1. Group all instructions by Primary Key (_id field) 
2. Convert each set of instructions into a MongoDB Write (or a few) 
3. Iterate through each primary key and apply the write to MongoDB (in a BulkWrite)

The issue is that the order of iteration for step 3 is non-deterministic. So you should observe that half the time your scenario works properly and half the time it results in a unique index constraint violation because the update is being applied before the delete. 

We have just filed a ticket we will try to get to in the next few weeks to make the ordering of writes more deterministic and follow the ordering that each primary key is first seen in an upload message. 

Thanks for finding, 
Tyler

**Comment by geosebas** [comment]
Hi, 

Thanks for the explanation ! 

Can you keep me updated when it will be fixed ?

**Comment by tkaye407** [comment]
Yes, of course. I would link you to a ticket but our Jira / Github are currently private. I have included a note in the task to ping this thread when it is completed and deployed. We are deploying a new version tomorrow, so I will try my best to get this into the following release (which is 2 weeks later). 

Please let me know if that works for you. If this is a production-related issue, we can try to backport this once it is completed, but if this is just an issue you came across in development we will hold off until the next release ideally. 

Thanks, 
Tyler

**Comment by tkaye407** [comment]
Hi, just a quick update that the issue was fixed and will be included in the release next Wednesday (March 29th). I will follow up on this ticket when it happens though. 

Thanks, 
Tyler

**Comment by tkaye407** [comment]
Hi, this was released yesterday. Please let me know if everything is working on your end and we can close this issue out! 

Thank you again for finding this issue, 
Tyler
